### PR TITLE
Include array classes in GetClassLoaderClasses

### PR DESCRIPTION
**Pull request by gacholio**
GetClassLoaderClasses returns the list of classes whose loading was
initiated by the given class loader, including the base type primitive
array classes. All array classes of these classes must also be returned.

Fixes: #9965

[ci skip]

Signed-off-by: Graham Chapman <graham_chapman@ca.ibm.com>

### CODE REVIEW

**Comment by gacholio** [comment]
@pshipton This could be considered for 0.21.

**Comment by gacholio** [comment]
jenkins test sanity,extended zlinux jdk8

**Comment by gacholio** [comment]
jenkins compile win jdk8

**Comment by gacholio** [comment]
jenkins test sanity,extended zlinux jdk8

**Comment by gacholio** [comment]
jenkins compile win jdk8

**Comment by gacholio** [comment]
Marking WIP while I investigate the visibilty of array classes between class loaders.

**Comment by gacholio** [comment]
Based on testing, it seems that GetClassLoaderClasses returns all of the classes in the hash table and all of their array classes along with all of the primitive array classes.

In my test, I have a class A (ClassLoader A). Within class A, a single and double-dimension array of A are created. Class B (ClassLoader B) extends A and does nothing. The results of GetClassLoaderClasses on each is:
```
A
	Ljava/lang/Object;
	[Ljava/lang/Object;
	[[Ljava/lang/Object;
	LA;
	[LA;
	[[LA;
	[Z
	[B
	[C
	[I
	[S
	[J
	[F
	[D
B
	LB;
	LA;
	[LA;
	[[LA;
	[Z
	[B
	[C
	[I
	[S
	[J
	[F
	[D
```

**Comment by gacholio** [comment]
With my current change, OpenJ9 returns:
```
A
	[J
	[[[I
	[[I
	[I
	[S
	[B
	[D
	[F
	[[C
	[C
	[Z
	[[[LA;
	[[LA;
	[LA;
	LA;
	Ljava/lang/StackOverflowError;
	[Ljava/lang/Object;
	Ljava/lang/Object;
	Ljava/lang/Cloneable;
	Ljava/lang/OutOfMemoryError;
	Ljava/io/Serializable;
	Ljava/lang/reflect/InvocationTargetException;
B
	[J
	[[[I
	[[I
	[I
	[S
	[B
	[D
	[F
	[[C
	[C
	[Z
	LB;
	[[[LA;
	[[LA;
	[LA;
	LA;
	Ljava/lang/StackOverflowError;
	Ljava/lang/OutOfMemoryError;
	Ljava/lang/reflect/InvocationTargetException;
```
Note that the API specifically says there is no ordering requirement.

**Comment by gacholio** [comment]
As an aside, it seems we're creating unnecessary array classes (note the triple dimension arrays of A and int). This has nothing to do with this PR.

**Comment by gacholio** [comment]
@DanHeidinga Ready for review.

**Comment by DanHeidinga** [comment]
@gacholio which version(s) of the RI did you compare with?

File: runtime/jvmti/jvmtiClass.c [hunk 1]
```diff
@@ -1367,15 +1381,20 @@ copyInitiatedClass(J9Class * clazz, J9JVMTIClassStats * results)
...
-		}
+
+		do {
+			jint slot = (jint)results->classCount - 1;
+
+			/* Reverse fill */
+			if (slot >= 0) {
+				j9object_t classObject = J9VM_J9CLASS_TO_HEAPCLASS(clazz);
+				
+				results->classRefs[slot] = (jclass)vm->internalVMFunctions->j9jni_createLocalRef(jniEnv, classObject);
```
Comment (DanHeidinga) [inline runtime/jvmti/jvmtiClass.c:1392]: Can you cache the `vm->internalVMFunctions` table in a local?  A minor thing but might be a small benefit for deep array chains...

**Comment by gacholio** [comment]
A random JDK8 - I don't believe the spec has changed.

File: runtime/jvmti/jvmtiClass.c [hunk 2]
```diff
@@ -237,39 +239,48 @@ jvmtiGetClassLoaderClasses(jvmtiEnv* env,
...
 			while (clazz != NULL) {
 				copyInitiatedClass(clazz, &stats);
-				clazz = vm->internalVMFunctions->hashClassTableNextDo(&walkState);
+				clazz = vmFuncs->hashClassTableNextDo(&hashWalkState);
 			}
+
+			/* The base type primitive array classes must also be added */
+			primitiveArray = &vm->booleanArrayClass;
+			do {
+				copyInitiatedClass(*primitiveArray++, &stats);
```
Comment (DanHeidinga) [inline runtime/jvmti/jvmtiClass.c:276]: I think we've been leaning towards retiring the use of `++` for non-primitive indexes and preferring an explicit ` += 1`.

File: runtime/jvmti/jvmtiClass.c (hunk 1 above)
Comment (gacholio) [reply runtime/jvmti/jvmtiClass.c:1392]: Done.

File: runtime/jvmti/jvmtiClass.c (hunk 2 above)
Comment (gacholio) [reply runtime/jvmti/jvmtiClass.c]: Done.

**Comment by DanHeidinga** [comment]
Found https://bugs.openjdk.java.net/browse/JDK-4450091 when looking at the spec which states the RI has the same behaviour of forcing all primitive arrays to be included.

**Comment by DanHeidinga** [comment]
Jenkins test sanity zlinux jdk8

**Comment by DanHeidinga** [comment]
Jenkins test extended osx jdk11

**Pull request merged**